Resolve strict dependencies from Ivy repositories

melix opened the pull request:
See #3462

bigdaz commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/IvyResolver.java:
```diff
@@ -153,6 +162,24 @@ protected MutableIvyModuleResolveMetadata parseMetaDataFromResource(ModuleCompon
+                moduleMetadataParser.parse(resource, metadata);
+            }
+        }
+        return metadata;
```
:x: This implementation is nearly identical to the one in `MavenResolver`. Should pull this up into superclass.
In fact, I think the current superclass implementation is now unused.

bigdaz commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/IvyResourcePattern.java:
```diff
@@ -51,10 +52,32 @@ public ExternalResourceName toVersionListPattern(ModuleIdentifier module, IvyArt
     }
 
     public ExternalResourceName toModulePath(ModuleIdentifier module) {
-        throw new UnsupportedOperationException("not implemented yet.");
```
This new implementation is incorrect: it should return the _module_ path (that contains all the version directories). However, since this isn't required for an Ivy resolver, we can just leave it with `UnsupportedOperationException`.

melix replied:
fixed

bigdaz commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/IvyResourcePattern.java:
```diff
@@ -51,10 +52,32 @@ public ExternalResourceName toVersionListPattern(ModuleIdentifier module, IvyArt
 
     public ExternalResourceName toModuleVersionPath(ModuleComponentIdentifier componentIdentifier) {
-        throw new UnsupportedOperationException("not implemented yet.");
+        ImmutableMap<String, String> attributes = ImmutableMap.of(
```
This implementation deserves a unit test.

melix replied:
done

bigdaz commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/MavenResolver.java:
```diff
@@ -156,6 +156,7 @@ private void resolveUniqueSnapshotDependency(MavenUniqueSnapshotComponentIdentif
         resolveStaticDependency(module, prescribedMetaData, result, createArtifactResolver(snapshotSource));
     }
 
+    // TODO:DAZ Pull this up
```
:x: I'll leave this to you :)

bigdaz commented on subprojects/dependency-management/src/main/java/org/gradle/internal/component/external/model/IvyDependencyMetadata.java:
```diff
@@ -101,6 +102,15 @@ public String getDynamicConstraintVersion() {
+            if (!consumerAttributes.isEmpty() || targetComponent instanceof IvyModuleResolveMetadata) {
+                return ImmutableSet.of(selectConfigurationUsingAttributeMatching(consumerAttributes, targetComponent, consumerSchema));
+            }
+        }
```
It would be great to pull this out so that we can avoid the duplication with `MavenDependencyMetadata`

melix replied:
I'm not sure how to handle this, and actually where it makes sense. The comment doesn't really help me either. I added this when chatting with @jjohannes , but I don't think we have code coverage for that. Any help in setting up such a test would be helpful.

jjohannes replied:
We should be able to test this by tests with transitive dependencies. And I think it will fail if you have 

`project --depends-on--> module-from-maven --depends-on--> module-from-ivy`

Because of the `targetComponent instanceof ...ModuleResolveMetadata`, which I _think_ should be removed from both Ivy and Maven. And then, the check `!targetComponent.getVariantsForGraphTraversal().isEmpty() && !consumerAttributes.isEmpty()` should go into the super class.

bigdaz replied:
@melix Will you look into adding some test coverage for this case? It would be great to fully understand the reason for this check.

It seems like original check for `instanceof MavenModuleResolveMetadata` was required before we added equivalent support for Ivy. At the very least, I think we should change the check to be `instanceof ModuleComponentResolveMetadata`, which will then be identical for Ivy and Maven.

melix replied:
Yes, this is on my todo list. But I need to understand how to reach this code first. Then refactoring is going to be easier.

bigdaz commented on subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/ivy/IvyFileModule.groovy:
```diff
@@ -261,6 +277,44 @@ class IvyFileModule extends AbstractModule implements IvyModule {
         return this
     }
 
+    private void publishModuleMetadata() {
```
:X: This duplication with `AbstractMavenModule` is a bit nasty. I want to add support for `excludes` and will have to do it in 2 places...

melix replied:
fixed

melix commented:
@bigdaz I have addressed all comments, except for the redundant code in `MavenDependencyMetadata`, which I think can be solved separately. Build passes. I will address `latest.status` in a subsequent PR, I need this one to make progress on test fixtures.

bigdaz commented on subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/GradleMetadataAwareExternalResourceResolver.java:
```diff
@@ -0,0 +1,75 @@
+
+import javax.annotation.Nullable;
+
+public abstract class GradleMetadataAwareExternalResourceResolver<T extends ModuleComponentResolveMetadata, S extends MutableModuleComponentResolveMetadata & MutableComponentVariantResolveMetadata> extends ExternalResourceResolver<T, S> {
```
I don't understand why we didn't just move this up to `ExternalResourceResolver`, since it's the current superclass of `IvyResolver` and `MavenResolver`.

melix replied:
I like to think of them as conceptually different (aware of metadata or not). For example there are subclasses of `ExternalResourceResolver` for tests which are _not_ aware of Gradle metadata.
